From f482657a186b0729c6af24ee58529a01d26d82fc Mon Sep 17 00:00:00 2001
From: Attila Gluck <gluck.attila.bela@gmail.com>
Date: Sun, 30 Aug 2020 18:44:37 +0200
Subject: [PATCH 2/2] dmenu-lineheight

---
 config.def.h |  2 ++
 dmenu.1      |  5 ++++-
 dmenu.c      | 19 +++++++++++++------
 3 files changed, 19 insertions(+), 7 deletions(-)

diff --git a/config.def.h b/config.def.h
index 1edb647..e9d80a7 100644
--- a/config.def.h
+++ b/config.def.h
@@ -15,6 +15,8 @@ static const char *colors[SchemeLast][2] = {
 };
 /* -l option; if nonzero, dmenu uses vertical list with given number of lines */
 static unsigned int lines      = 0;
+/* -h option; minimum height of a menu line     */
+static unsigned int lineheight = 0;
 
 /*
  * Characters not considered part of a word while deleting words
diff --git a/dmenu.1 b/dmenu.1
index 762f707..1806156 100644
--- a/dmenu.1
+++ b/dmenu.1
@@ -3,7 +3,7 @@
 dmenu \- dynamic menu
 .SH SYNOPSIS
 .B dmenu
-.RB [ \-bfivP ]
+.RB [ \-bfivPl ]
 .RB [ \-l
 .IR lines ]
 .RB [ \-m
@@ -50,6 +50,9 @@ dmenu matches menu items case insensitively.
 .B \-P
 dmenu will not directly display the keyboard input, but instead replace it with dots. All data from stdin will be ignored.
 .TP
+.BI \-h " height"
+dmenu uses a menu line of at least 'height' pixels tall, but no less than 8.
+.TP
 .BI \-l " lines"
 dmenu lists items vertically, with the given number of lines.
 .TP
diff --git a/dmenu.c b/dmenu.c
index 4a56df7..f9b8fa4 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -131,7 +131,7 @@ drawmenu(void)
 {
 	unsigned int curpos;
 	struct item *item;
-	int x = 0, y = 0, w;
+	int x = 0, y = 0, fh = drw->fonts->h, w;
     char *censort;
 
 	drw_setscheme(drw, scheme[SchemeNorm]);
@@ -154,8 +154,8 @@ drawmenu(void)
 	curpos = TEXTW(text) - TEXTW(&text[cursor]);
 	if ((curpos += lrpad / 2 - 1) < w) {
 		drw_setscheme(drw, scheme[SchemeNorm]);
-		drw_rect(drw, x + curpos, 2, 2, bh - 4, 1, 0);
-	}
+		drw_rect(drw, x + curpos, 2 + (bh-fh)/2, 2, fh - 4, 1, 0);
+    }
 
 	if (lines > 0) {
 		/* draw vertical list */
@@ -619,8 +619,10 @@ setup(void)
 	utf8 = XInternAtom(dpy, "UTF8_STRING", False);
 
 	/* calculate menu geometry */
-	bh = drw->fonts->h + 2;
-	lines = MAX(lines, 0);
+    bh = drw->fonts->h+2;
+    /* make a menu line AT LEAST 'lineheight' tall */
+	bh = MAX(bh,lineheight);
+    lines = MAX(lines, 0);
 	mh = (lines + 1) * bh;
 #ifdef XINERAMA
 	i = 0;
@@ -701,7 +703,8 @@ static void
 usage(void)
 {
 	fputs("usage: dmenu [-bfiPv] [-l lines] [-p prompt] [-fn font] [-m monitor]\n"
-	      "             [-nb color] [-nf color] [-sb color] [-sf color] [-w windowid]\n", stderr);
+	      "             [-h height] [-nb color] [-nf color] [-sb color] [-sf color]\n"
+          "             [-w windowid]\n", stderr);
 	exit(1);
 }
 
@@ -736,6 +739,10 @@ main(int argc, char *argv[])
 			prompt = argv[++i];
 		else if (!strcmp(argv[i], "-fn"))  /* font or font set */
 			fonts[0] = argv[++i];
+		else if(!strcmp(argv[i], "-h")) { /* minimum height of one menu line */
+			lineheight = atoi(argv[++i]);
+			lineheight = MAX(lineheight,8); /* reasonable default in case of value too small/negative */
+		}
 		else if (!strcmp(argv[i], "-nb"))  /* normal background color */
 			colors[SchemeNorm][ColBg] = argv[++i];
 		else if (!strcmp(argv[i], "-nf"))  /* normal foreground color */
-- 
2.28.0

